File: src/com/leo/leetcode/algorithm/Q97.java
package com.leo.leetcode.algorithm;

/**
 * 给定三个字符串 s1, s2, s3, 验证 s3 是否是由 s1 和 s2 交错组成的。
 * 链接：https://leetcode-cn.com/problems/interleaving-string/
 */
public class Q97 {
    public static void main(String[] args) {
        System.out.println(new Q97().isInterleave("aabcc", "dbbca", "aadbbcbcac")); // true
        System.out.println(new Q97().isInterleave("aabcc", "dbbca", "aadbbbaccc")); // false
        System.out.println(new Q97().isInterleave("", "", "")); // true
        System.out.println(new Q97().isInterleave("a", "", "a")); // true
    }

    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        boolean[][][] dp = new boolean[s3.length() + 1][s1.length() + 1][s2.length() + 1]; // s2, s3, s1
        for (int i = 0; i <= s1.length(); i++) for (int j = 0; j <= s2.length(); j++) dp[0][i][j] = false;
        dp[0][0][0] = true;
        for (int i = 1; i <= s3.length(); i++) {
            char c = s3.charAt(i - 1);
            for (int i1 = 0; i1 <= s1.length() && i1 <= i; i1++) {
                int i2 = i - i1;
                if (i2 > s2.length()) continue;
                if (i1 == 0) dp[i][i1][i2] = dp[i - 1][i1][i2 - 1] && s2.charAt(i2 - 1) == c;
                else if (i2 == 0) dp[i][i1][i2] = dp[i - 1][i1 - 1][i2] && s1.charAt(i1 - 1) == c;
                else {
                    dp[i][i1][i2] = (s1.charAt(i1 - 1) == c && dp[i - 1][i1 - 1][i2])
                            || (s2.charAt(i2 - 1) == c && dp[i - 1][i1][i2 - 1]);
                }
            }
        }
        return dp[s3.length()][s1.length()][s2.length()];

    }
}

File: py/q0000/Q97.py
"""
 * 给定三个字符串 s1, s2, s3, 验证 s3 是否是由 s1 和 s2 交错组成的。
 * 两个字符串 s 和 t 交错 的定义与过程如下，其中每个字符串都会被分割成若干 非空 子字符串：
 * s = s1 + s2 + ... + sn
 * t = t1 + t2 + ... + tm
 * |n - m| <= 1
 * 交错 是 s1 + t1 + s2 + t2 + s3 + t3 + ... 或者 t1 + s1 + t2 + s2 + t3 + s3 + ...
 * 注意：a + b 意味着字符串 a 和 b 连接。
 * 提示：
 * 1、0 <= s1.length, s2.length <= 100
 * 2、0 <= s3.length <= 200
 * 3、s1、s2、和 s3 都由小写英文字母组成
 * 进阶：您能否仅使用 O(s2.length) 额外的内存空间来解决它?
 * 链接：https://leetcode-cn.com/problems/interleaving-string/
"""
from functools import cache


class Solution:

    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        # dp 解决，dp[i][j] 表示 s1[:i] 和 s2[:j] 能否交错组成 s3[:i+j]
        l1, l2, l3 = len(s1), len(s2), len(s3)
        if l3 != l1 + l2:
            return False

        dp = [True] + [False] * (l2)
        for i in range(l1 + 1):
            for j in range(l2 + 1):
                if i:
                    dp[j] &= s1[i - 1] == s3[i + j - 1]
                if j and s2[j - 1] == s3[i + j - 1]:
                    dp[j] |= dp[j - 1]
        return dp[-1]

    def isInterleave1(self, s1: str, s2: str, s3: str) -> bool:
        if len(s3) != len(s1) + len(s2):
            return False

        @cache
        def dfs(idx, i1, i2):
            if idx == len(s3):
                return True
            ret = False
            if i1 < len(s1) and s3[idx] == s1[i1]:
                ret |= dfs(idx + 1, i1 + 1, i2)
            if i2 < len(s2) and s3[idx] == s2[i2]:
                ret |= dfs(idx + 1, i1, i2 + 1)
            return ret

        return dfs(0, 0, 0)


if __name__ == '__main__':
    # True
    print(Solution().isInterleave("aabcc", s2="dbbca", s3="aadbbcbcac"))
    # False
    print(Solution().isInterleave("aabcc", s2="dbbca", s3="aadbbbaccc"))
    # True
    print(Solution().isInterleave("", s2="", s3=""))
